crypto/bn/random.c

```diff
@@ -118,120 +118,21 @@
 #include "../internal.h"
 
 
-int BN_rand(BIGNUM *rnd, int bits, int top, int bottom, RAND *rng) {
-  uint8_t *buf = NULL;
-  int ret = 0, bit, bytes, mask;
-
-  if (rnd == NULL) {
-    return 0;
-  }
-
-  if (bits == 0) {
-    BN_zero(rnd);
-    return 1;
-  }
-
-  bytes = (bits + 7) / 8;
-  bit = (bits - 1) % 8;
-  mask = 0xff << (bit + 1);
-
-  buf = OPENSSL_malloc(bytes);
-  if (buf == NULL) {
-    OPENSSL_PUT_ERROR(BN, ERR_R_MALLOC_FAILURE);
-    goto err;
-  }
-
-  /* Make a random number and set the top and bottom bits. */
-  if (!RAND_bytes(rng, buf, bytes)) {
-    goto err;
-  }
-
-  if (top != -1) {
-    if (top && bits > 1) {
-      if (bit == 0) {
-        buf[0] = 1;
-        buf[1] |= 0x80;
-      } else {
-        buf[0] |= (3 << (bit - 1));
-      }
-    } else {
-      buf[0] |= (1 << bit);
-    }
-  }
-
-  buf[0] &= ~mask;
-
-  /* set bottom bit if requested */
-  if (bottom)  {
-    buf[bytes - 1] |= 1;
-  }
-
-  if (!BN_bin2bn(buf, bytes, rnd)) {
-    goto err;
-  }
-
-  ret = 1;
-
-err:
-  OPENSSL_free(buf);
-  return (ret);
-}
-
 int BN_rand_range(BIGNUM *r, const BIGNUM *range, RAND *rng) {
-  unsigned n;
-  unsigned count = 100;
-
   if (range->neg || BN_is_zero(range)) {
     OPENSSL_PUT_ERROR(BN, BN_R_INVALID_RANGE);
     return 0;
   }
 
-  n = BN_num_bits(range); /* n > 0 */
-
-  /* BN_is_bit_set(range, n - 1) always holds */
-  if (n == 1) {
-    BN_zero(r);
-  } else if (!BN_is_bit_set(range, n - 2) && !BN_is_bit_set(range, n - 3)) {
-    /* range = 100..._2,
-     * so  3*range (= 11..._2)  is exactly one bit longer than  range */
-    do {
-      if (!BN_rand(r, n + 1, -1 /* don't set most significant bits */,
-                   0 /* don't set least significant bits */, rng)) {
-        return 0;
-      }
-
-      /* If r < 3*range, use r := r MOD range (which is either r, r - range, or
-       * r - 2*range). Otherwise, iterate again. Since 3*range = 11..._2, each
-       * iteration succeeds with probability >= .75. */
-      if (BN_cmp(r, range) >= 0) {
-        if (!BN_sub(r, r, range)) {
-          return 0;
-        }
-        if (BN_cmp(r, range) >= 0) {
-          if (!BN_sub(r, r, range)) {
-            return 0;
-          }
-        }
-      }
-
-      if (!--count) {
-        OPENSSL_PUT_ERROR(BN, BN_R_TOO_MANY_ITERATIONS);
-        return 0;
-      }
-    } while (BN_cmp(r, range) >= 0);
-  } else {
-    do {
-      /* range = 11..._2  or  range = 101..._2 */
-      if (!BN_rand(r, n, -1, 0, rng)) {
-        return 0;
-      }
+  if (bn_wexpand(r, range->top) == NULL) {
+    return 0;
+  }
 
-      if (!--count) {
-        OPENSSL_PUT_ERROR(BN, BN_R_TOO_MANY_ITERATIONS);
-        return 0;
-      }
-    } while (BN_cmp(r, range) >= 0);
+  if (!GFp_rand_mod(r->d, range->d, range->top, rng)) {
+    return 0;
   }
 
+  r->top = range->top;
```

Comment:
This needs to be a call to `bn_correct_top`. `BIGNUM`s always need to have the minimal value of `top` so that the top-level limb is never zero. (This destroys the constant-time property of `BIGNUM`, which is why we're getting rid of it. But certain other code depends on it, so whenever we have a `BIGNUM` we have to maintain that invariant.)